### PR TITLE
(DRAFT) Issue #3 - Database Redesign

### PR DESCRIPTION
**Pull request by farshidgoudarzi**
Issue https://github.com/farshidgoudarzi/messenger-2670/issues/3:

To allow the creation of conversations with 2 or more than 2 users, I decided to create a middle table called "UserConversations". So basically, a user can have many UserConversations, and each UserConversation has a foreign key to a User and Conversation, and each Conversation can have many UserConversations. This way we allow the conversation to have as many participants as requested.

Based on how we had addressed the read flag for each message in the previous issue, now just the Boolean flag is no longer enough (a message can have more than 1 viewer), so to solve this, I created another table called "MessageUserInteractions", this has the same principle as the other, a message can have multiple MessageUserInteractions and that has one user. So we can have multiple user interactions (read flags) for each message based on who has viewed the message.

Below I have also attached my database mock up document.

To answer the questions listed in the issue:
**Question 1:**
Some of the additional changes would include:
- Changes to the design and UI to include creating chats with more than one participants
- Changes to the design and UI to include chats messages with multiple participants (Who are in the chat, who viewed the message, ...)
- Changes to creating chat flow and marking read flag for message flow, based on what is described above.
- Database and API migration to v2.

**Question 2:**
We can create a simple data migration to move the users data from the previous structure to new structure. For conversations, instead of user1 and 2, we create records in the UserConversations table and instead of the message read flag we create record in MessageUserInteractions table (and any other changes to the data that might be necessary). This will be a solution to moving data from previous version without risking of compromising the users data, however this will still cause momentary disruption for users using the app in the time of deployment.

To prevent that (in case based on the project resources it would make sense), we can create a duplicate instance of the database (v2 for example), new version client would use new instance, old version would use the older instance, and we reflect any changes coming to old version on the new instance (using triggers, etc...). This way whenever all the current clients are moved to the new version (if app, update, if web, next use) and only then, old instance can be taken out of order. This way there is no disruption at all.

[Issue #3 Data Structure Solution.pdf](https://github.com/farshidgoudarzi/messenger-2670/files/7311968/Issue.3.Data.Structure.Solution.pdf)

### CODE REVIEW

File: server/db/models/index.js [hunk 1]
```diff
@@ -1,14 +1,22 @@
...
+UserConversation.belongsTo(User);
+
+Conversation.hasMany(UserConversation);
+UserConversation.belongsTo(Conversation);
+
 Conversation.hasMany(Message);
+Message.belongsTo(Conversation);
+
+Message.hasMany(MessageUserInteraction);
+MessageUserInteraction.belongsTo(Message);
```
Comment (FilipeGorodscy) [inline server/db/models/index.js:19]: You can leverage the Sequelize's `belongsToMany` method

**Review by FilipeGorodscy — COMMENTED** [review]
Awesome work! I have some other questions:
- How would you update route handlers to reflect the schema change?
- How would you update socket event handlers to reflect the schema change?
- How would you update the Redux state to reflect the schema change?
- If we have multiple developers involved in this project, how can we split the task between them without introducing merge conflict?

Please let us know on Slack when this PR is ready for another review.

**Comment by farshidgoudarzi** [comment]
**Q1:** 
- Based on what we have right now, we can keep the read-flag under conversation because I see at as a user marks a conversation as read, therefore the messages of the conversation are all read. But if we think of it as the user, can mark certain messages as read and not all (like if there are a lot of new messages and user only sees certain number at a time) we can move the mark-read under messages.
- For getting data, conversation should return list of users instead of user1 and 2.
- Messages would return a list of interactions instead of the read flag for each message (which user has read the message). We can map this into a list of userIds that have read the message, if we have only read flag as interactions.

**Q2:** 
- Adding and removing the online/offline users should change. Each user in the conversation can have an online/offline status now.
- Read flag should only be valid for the message sender.
- Logic for new message can stay the same if we keep the logic on the Conversation level. (Might need small tweaks) 

**Q3:** 
- Active conversation can no longer be other users username, I would change it to the conversation Id.
- Conversations state and actions should be updated according to the new changes (otherUser is irrelevant, we have a list of participants or users for the conversation everywhere.)
- Lot of other small details about each of the actions ... (let me know if I should explain more details)

**Q4:** 
- I would separate backend and front end. Backend developers can create a mock up of the new data structure for the front end team to be able to work with the data and create the front end, while backend developers make the changes to routes and database. 

Let me know if I should dig deeper.

**Review by FilipeGorodscy — APPROVED** [review]
Looks good!
- Instead of an array/list of other users, is there any other data structure that could improve efficiency on lookups/add/delete?

No need to merge this PR 👍

**Comment by farshidgoudarzi** [comment]
Yes, Maps (Objects) would be better. With keys being the userId, and values can be the data needed. With only the read-flag, we value can be Boolean value of the read-flag, otherwise value can be the Object of the data we need. This way we can lookup or update based on key which is a lot more efficient.

example with only read flag:
```
{ 
  1: true,
  2: false,
}
```
example with interaction data:
```
{ 
  7: {
    isRead: true,
    someOtherInteraction: false
  },
  8: {
    isRead: false,
    someOtherInteraction: false
  },
}
```

**Pull request closed without merging**